### DISCUSSION TITLE
Uasyncio StreamReader socket and uart togheter, data loss.

### DISCUSSION THREAD

**Discussion by prem111as** [micropython/micropython#14185] [General]
Hello. I am using two StreamReader streams, one for UART the other for Socket udp. If there is a lot of data from UART, unfortunately I lose UDP packets. Is there any solution for this?

Regards.

**Comment by sophiedegran** [comment]
Hi
Did you look at auart.py from Peter on github.com/peterhinch/micropython-async/v3/as_demos ?

**Reply by prem111as** [reply]
Yes, that's exactly what I do.

**Reply by sophiedegran** [reply]
So reading uart is not the problem ... and if I understand you have among of data to send to udp and you loose them?

**Reply by prem111as** [reply]
Yes but I just only receive udp data and lose it. 
code:

`
import usocket as socket

sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM, socket.IPPROTO_UDP)
sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)

sock.bind(('', MCAST_PORT))

sreader = uasyncio.stream.StreamReader(sock)

while True:
    recv = await sreader.read(20)
    if (recv):
        print(recv)
    await uasyncio.sleep(0)
`

**Reply by sophiedegran** [reply]
increase read(20) ... despite I don't know how is your server... and if he send the good udp data after reading uart  (overflow?)

**Reply by prem111as** [reply]
I only need 20 first bytes, some lines arrive correctly, some are completely lost. This only happens when I receive a lot of data via UART at the same time.

**Reply by sophiedegran** [reply]
Ok so I think you have to Queue before sending on your server.. you will loose a little the 'real time' but that will be ok

**Reply by prem111as** [reply]
I cannot use the queue, I receive UDP data from a physical device (which I cannot configure) via WiFi.

**Reply by sophiedegran** [reply]
Nn by Queuing I would say buffering ☺ 
Look at Peter examples on primitives queue.py and ringbuf_queue.py

**Reply by prem111as** [reply]
Can I find examples of using these classes somewhere? I have one more question, when I skip StreamReader and read directly from sock.recv() it gets all the data. Unfortunately the loop hangs when the udp device stops responding, can I use something to prevent blocking?

**Reply by sophiedegran** [reply]
Yes look at the TUTORIAL.md in /docs it look a little 'hard' to use primitives from Peter but it's not...
Take your time you find the problem
1- async read uart
2- async queue
3- async serve udp

**Reply by prem111as** [reply]
Thanks ! Can you tell me something about my second question about socket and blocking ?

**Reply by bixb922** [reply]
If UDP blocks for a longer time, asyncio cannot serve the UART queue and it will overflow, and characters will be lost.

Some options:
Option 1. Use non blocking sockets and check the socket for example every 20 milliseconds waiting with asyncio.sleep_ms. This is the simplest option, except if the UDP messages were very, very time critical

Option 2. Start a _thread for the code related to the socket. Coordination with the asyncio code for the UART can be a bit tricky, by all means see https://github.com/peterhinch/micropython-async/blob/master/v3/docs/THREADING.md

Option 3. There may be some asyncio UDP library out there.

**Comment by prem111as** [comment]
There is one more thing. Why is it that when I receive a socket stream via StreamReader, when I disconnect the device that serves data via udp, UART reading does not block, and when I getting data directly from sock.recv, it blocks when disconnect device. s there any way to prevent sock.recv from blocking?

**Reply by sophiedegran** [reply]
Is that the code you send above? If so make an exception in your while..

**Reply by prem111as** [reply]
Yes its it. What exactly do you mean about exception in while ? can you give an example? in my case, when the device that serves UDP is disconnected, the loop just hangs and is blocked on sock.recv..... I know I can use settimeout, but I would like to avoid it ( I would like this async task to be independent of what happens in the sock, whether it is blocked or not )

**Reply by sophiedegran** [reply]
If I understand you il the server stop sending udp it's hang on your client (code above) and you should have an exception on it?

**Reply by prem111as** [reply]
Yes, exactly. This is what happens, when the udp server stops sending, my sock.revc loop hangs, it blocks, and I don't want that, I don't want blocking.

**Comment by sophiedegran** [comment]
Look at bixb answer.. (THREADING.md)
For me:
1- put the udp receiver in async mode (it is not)
2- try to have an exception in your context (should not block)

**Reply by prem111as** [reply]
This is what my code looks like when I disconnect the device that serves udp, the script hangs on sock.recv


`

async def udp_service(self):

            try:
                import usocket as socket
                MCAST_GRP = '224.0.0.1'
                MCAST_PORT = 4444
                sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM, socket.IPPROTO_UDP)
                sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
                sock.bind(('', MCAST_PORT))
                while True:
                    recv = sock.recv(20)
                    print(recv)
                    await uasyncio.sleep(0)
            except Exception as e:
                print("Exception: "+str(e))`

**Reply by sophiedegran** [reply]
You are listening a broadcast adress.. 
Please google micropython udp

**Reply by prem111as** [reply]
Yes, broadcast, it is correct, question is what can I do with my code so that it does not block reading when the device is turned off it stops serving data.

**Reply by peterhinch** [reply]
My approach would be that recommended by @bixb922 in option 1: use a nonblocking socket. Some example code is [here](https://github.com/peterhinch/micropython-mqtt/blob/d74121c49021da645ef09844bce00ab111521086/mqtt_as/mqtt_as.py#L202). The `._as_read` method does a nonblocking read of `n` characters, handling errors including a timeout.

**Reply by prem111as** [reply]
But I have no influence on the device that sends UDP data. I'm only receiving data via UDP.

**Reply by bixb922** [reply]
Yes, that's OK, one cannot control when data will arrive.
Use sock.setblocking(False), which is the same as sock.settimeout(0), to make the socket nonblocking. With that option, sock.recv() will never block.

At what average rate are you expecting the UDP messages to arrive (messages per second)?

**Reply by prem111as** [reply]
When use sock.setblocking(False) i given error: [Errno 11] EAGAIN

~10 messages per second.

**Reply by sophiedegran** [reply]
You're listening all broadcast udp ! Listen only your server..
You receive all dns call, etc...

**Reply by prem111as** [reply]
Server sent multicast messages, but even when I specify sock.bind(('192.168.0.1', 4444)) and setblocking(False) i given the same error: [Errno 11] EAGAIN

**Reply by sophiedegran** [reply]
Why you do that? Do a connexion point to point ..
Server should listening on 4444 and client reading the 4444 of the server
On broadcast you have many things.. dns, discovery.. 
Perharps put your server in AP mode and try with the client connected on it.. and increase the socket.recv () still you have not resolve that

**Reply by prem111as** [reply]
I don't understand, the server sends multicast messages to a specific port, how should I connect point to point? Can you provide example code, what do you mean?

**Reply by bixb922** [reply]
> When use sock.setblocking(False) i given error: [Errno 11] EAGAIN
> 
> ~10 messages per second.

Yes, that means that there is no data when the receive is executed. Trap the exception and await asyncio.sleep_ms(something) between each retry
